### PR TITLE
First implementation of terraform-aws-alb-single-listener

### PR DESCRIPTION
**Pull request by salvianreynaldi**

### CODE REVIEW

File: outputs.tf [hunk 1]
```diff
@@ -0,0 +1,7 @@
+output "lb-fqdn" {
+  value = "${aws_route53_record.main.fqdn}"
```
Comment (febryantonius) [inline outputs.tf]: please add description

File: variables.tf [hunk 2]
```diff
@@ -0,0 +1,138 @@
+variable "lb_name" {
+  default     = ""
+  description = ""
```
Comment (febryantonius) [inline variables.tf]: please fill the description

File: variables.tf [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+    "type"            = "lb_cookie"
+    "cookie_duration" = 1
+    "enabled"         = true
+  }
+
+  description = ""
+}
+
+variable "tg_name" {
+  default     = ""
```
Comment (febryantonius) [inline variables.tf]: can target group name be empty?

Comment (salvianreynaldi) [reply variables.tf]: `tg_name` and `lb_name` is used to override the default name (see **locals.tf** above)
I think override naming is useful when importing existing resources into the module

File: README.md [hunk 4]
```diff
@@ -1,2 +1,8 @@
 # terraform-aws-route53-alb_single_listener
-A terraform module which provisions a DNS record that points to an Application LB that has a single listener
+A terraform module which provisions a DNS record that points to an Application LB that has a single listener. Listener rules might be added along with a list of target group arns and a listener-target_group mapping.
+
+
+# Conventions
+- the listener use port `443`, `HTTPS` protocol, and `ELBSecurityPolicy-2016-08`
+- taggable resources will have `Name`, `Service`, `Environment`, and `Description` tags; their value can be overriden
+- specify 0 in `listener_target_group_idx` to select the default target group
```
Comment (febryantonius) [inline README.md]: add license

File: examples/alb-plus-backend-and-frontend-tgs/main.tf [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+      "values" = ["flight.traveloka.com"]
+    },
+    {
+      "field"  = "path-pattern"
+      "values" = ["/*/g"]
+    },
+  ]
+}
+
+variable "black-panther-condition" {
```
Comment (abihf) [inline examples/alb-plus-backend-and-frontend-tgs/main.tf]: ?

File: main.tf [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+  security_groups = ["${var.lb_security_groups}"]
+
+  subnets         = "${var.lb_subnet_ids}"
+  idle_timeout    = "${var.lb_idle_timeout}"
+  ip_address_type = "${var.lb_ip_address_type}"
+
+  access_logs {
+    bucket  = "${var.lb_logs_s3_bucket_name}"
+    prefix  = "${local.lb_name}"
+    enabled = true
```
Comment (abihf) [inline main.tf]: please add variable to configure it

File: main.tf [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+    "Environment", var.tag_environment,
+    "Description", var.tag_description,
+    ), var.lb_tags)}"
+}
+
+resource "aws_lb_listener" "main" {
+  load_balancer_arn = "${aws_lb.main.arn}"
+  port              = 443
+  protocol          = "HTTPS"
+  ssl_policy        = "ELBSecurityPolicy-2016-08"
```
Comment (abihf) [inline main.tf]: please add variable to configure it

File: outputs.tf [hunk 8]
```diff
@@ -0,0 +1,9 @@
+output "lb-fqdn" {
+  value       = "${aws_route53_record.main.fqdn}"
+  description = "the FQDN of the DNS record that points to the ALB"
+}
+
+output "default-target-group-arn" {
+  value       = "${aws_lb_target_group.default.arn}"
+  description = "the arn of the default target group"
+}
```
Comment (abihf) [inline outputs.tf]: please also output ALB's arn and arn_suffix (for autoscaling group policy)

File: main.tf [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+}
+
+resource "aws_lb_target_group" "default" {
+  name                 = "${local.tg_name}"
+  port                 = "${var.tg_port}"
+  protocol             = "${var.tg_protocol}"
+  vpc_id               = "${var.vpc_id}"
+  deregistration_delay = "${var.tg_deregistration_delay}"
+
+  health_check = ["${var.tg_healthcheck}"]
```
Comment (abihf) [inline main.tf]: can we merge this to default health check config? 
for example:
```hcl
health_check = ["${merge(locals.default_tg_healthchek, var.tg_healthcheck)}"]
```

```hcl
# locals.tf
locals {
  default_tg_healthcheck = {
    "interval"            = 30
    "path"                = "/healthcheck"
    "port"                = 8080
    "healthy_threshold"   = 3
    "unhealthy_threshold" = 3
    "timeout"             = 5
    "protocol"            = "HTTP"
    "matcher"             = "200"
  }
}
```

File: README.md [hunk 10]
```diff
@@ -1,2 +1,8 @@
 # terraform-aws-route53-alb_single_listener
```
Comment (andysaputra) [inline README.md]: IMO, the naming can be misleading. The main resource should be the ALB, not the route53, right?

File: variables.tf (hunk 3 above)
Comment (salvianreynaldi) [reply variables.tf]: turns out importing resources to module is only possible if we use only variables in resources' `name` parameter, i.e. locals and interpolation syntax will trigger resource recreation
so this is now here only to allow name overriding

File: main.tf [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+  protocol             = "${var.tg_protocol}"
+  vpc_id               = "${var.vpc_id}"
+  deregistration_delay = "${var.tg_deregistration_delay}"
+
+  health_check = ["${local.tg_health_check}"]
+
+  stickiness = ["${var.tg_stickiness}"]
+
+  tags = "${merge(map(
+    "Name", local.tg_name,
```
Comment (andysaputra) [inline main.tf]: Missing ProductDomain

File: outputs.tf [hunk 12]
```diff
@@ -0,0 +1,24 @@
+output "lb-fqdn" {
+  value       = "${aws_route53_record.main.fqdn}"
+  description = "the FQDN of the DNS record that points to the ALB"
+}
+
+output "lb-arn" {
+  value       = "${aws_lb.main.arn}"
+  description = "the FQDN of the DNS record that points to the ALB"
```
Comment (andysaputra) [inline outputs.tf]: wrong description

File: outputs.tf [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+}
+
+output "lb-arn-suffix" {
+  value       = "${aws_lb.main.arn_suffix}"
+  description = "the FQDN of the DNS record that points to the ALB"
```
Comment (andysaputra) [inline outputs.tf]: same as above

File: outputs.tf [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+}
+
+output "tg-arn" {
+  value       = "${aws_lb_target_group.default.arn}"
+  description = "the arn of the default target group"
+}
+
+output "tg-arn-suffix" {
+  value       = "${aws_lb_target_group.default.arn_suffix}"
+  description = "the arn of the default target group"
```
Comment (andysaputra) [inline outputs.tf]: wrong?

File: examples/frontend-backend-and-default-tgs/main.tf [hunk 15]
```diff
@@ -0,0 +1,119 @@
...
+  }
+
+  tags = "${map(
+    "Name", "fpr-backend-canary",
+    "Service", "fprbe",
+    "Environment", "production",
+  )}"
+}
+
+module "alb-route53" {
```
Comment (andysaputra) [inline examples/frontend-backend-and-default-tgs/main.tf]: maybe change to a more realistic example

File: locals.tf [hunk 16]
```diff
@@ -0,0 +1,26 @@
+resource "random_string" "lb_suffix" {
```
Comment (andysaputra) [inline locals.tf]: i think random_id is preferred

File: README.md [hunk 17]
```diff
@@ -1,2 +1,12 @@
-# terraform-aws-route53-alb_single_listener
-A terraform module which provisions a DNS record that points to an Application LB that has a single listener
+# terraform-aws-alb-single-listener-and-route53-record
```
Comment (andysaputra) [inline README.md]: I start to wonder whether creating the Route 53 within the module is a good idea or not.

Will there be a use case in which we want to do DNS routing (Blue Green migration) ?

File: main.tf (hunk 6 above)
Comment (abihf) [reply main.tf]: no, but it's public module

Comment (salvianreynaldi) [reply main.tf]: I think we're also applying some internal conventions (Tags, Naming) on module-level, so this should be enforced too

File: outputs.tf [hunk 18]
```diff
@@ -0,0 +1,29 @@
+output "lb-dns" {
```
Comment (andysaputra) [inline outputs.tf]: i think the convention is to use underscore instead of hyphen?

**Pull request merged**